src/Extensions/WakaTime.cpp

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2019-2021 Ashar Khan <ashar786khan@gmail.com>
+ *
+ * This file is part of CP Editor.
+ *
+ * CP Editor is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * I will not be responsible if CP Editor behaves in unexpected way and
+ * causes your ratings to go down and or lose any important contest.
+ *
+ * Believe Software is "Software" and it isn't immune to bugs.
+ *
+ */
+
+#include "Extensions/WakaTime.hpp"
+#include "Core/EventLogger.hpp"
+#include <QProcess>
+
+namespace Extensions
+{
+
+WakaTime::WakaTime(const QString &path, const QString &key, QObject *parent)
+    : QObject(parent), wakaTimePath(path), apiKey(key)
+{
+    LOG_INFO(INFO_OF(path));
+    LOG_INFO(INFO_OF(key));
+}
+
+void WakaTime::update(const QString &path, const QString &key)
+{
+    wakaTimePath = path;
+    apiKey = key;
+}
+
+void WakaTime::sendHeartBeat(const QString &filePath, bool isWrite)
+{
+    QTime now = QTime::currentTime();
+    if (!isWrite && filePath == lastFilePath && lastHeartBeat.isValid() && lastHeartBeat.msecsTo(now) > 2 * 60 * 1000)
+        return;
+    lastHeartBeat = now;
+    lastFilePath = filePath;
+    auto *wakaTimeProcess = new QProcess();
```

Comment:
Add error handling and more logs here, like if process failed to start, the exit code it returned and anything else. In case of failure to start/non zero exit you can show a error in messagelogger to user.

Reply:
> In case of failure to start/non zero exit you can show a error in messagelogger to user.

That's the most terrible thing when I'm using NeoVim under a bad network.

Reply:
I think no one cares if WakaTime doesn't work for a single time. And they can notice that there are no stats on the WakaTime dashboard if it always doesn't work.